### PR TITLE
update deployment and replicaset listers

### PR DESCRIPTION
**Pull request by deads2k**
Updates the deployment lister to avoid copies and updates the deployment controller to use shared informers.

Pushing WIP to see which tests are broken.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34024)

### CODE REVIEW

**Comment by deads2k** [comment]
@derekwaynecarr This got unexpectedly large.  Let's start with this and I'll do the informer in another step.

**Comment by deads2k** [comment]
@ncdc you got review bandwidth?  We will be using these downstream eventually (backing a DC by an RS is needed to help a migration) and this is also groundwork for fixing up more duplicate listwatches.

**Comment by ncdc** [comment]
Yes, in an hour or so

File: pkg/client/cache/listers_extensions.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// Example:
+// s := cache.NewStore()
+// lw := cache.ListWatch{Client: c, FieldSelector: sel, Resource: "pods"}
+// r := cache.NewReflector(lw, &extensions.Deployment{}, s).Run()
+// l := StoreToDeploymentLister{s}
+// l.List()
+
+// StoreToDeploymentLister helps list pods
+
+// StoreToDeploymentLister gives a store List and Exists methods. The store must contain only Deployments.
```
Comment (ncdc) [inline pkg/client/cache/listers_extensions.go]: You didn't migrate Exists. Change comment to only mention List?

File: pkg/client/cache/listers_extensions.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	err = ListAll(s.Indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(*extensions.Deployment))
+	})
+	return ret, err
+}
+
+func (s *StoreToDeploymentLister) Deployments(namespace string) storeDeploymentsNamespacer {
+	return storeDeploymentsNamespacer{Indexer: s.Indexer, namespace: namespace}
+}
+
```
Comment (ncdc) [inline pkg/client/cache/listers_extensions.go]: There is some godoc for this type here if you want it: https://github.com/kubernetes/kubernetes/pull/34024/commits/25f42a23721dec67af3492eba087170bd1993bb8#diff-53f28720440ab4b589ef2273346cecfcL197

File: pkg/client/cache/listers_extensions.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		err = fmt.Errorf("could not find deployments set for ReplicaSet %s in namespace %s with labels: %v", rs.Name, rs.Namespace, rs.Labels)
+	}
+	return
+}
+
+// GetDeploymentsForDeployments returns a list of deployments managing a pod. Returns an error only if no matching deployments are found.
+// TODO eliminate shallow copies
+func (s *StoreToDeploymentLister) GetDeploymentsForPod(pod *api.Pod) (deployments []*extensions.Deployment, err error) {
+	if len(pod.Labels) == 0 {
+		err = fmt.Errorf("no deployments found for Deployment %v because it has no labels", pod.Name)
```
Comment (ncdc) [inline pkg/client/cache/listers_extensions.go]: The original error was "no deployments found for Pod" - change it back?

File: pkg/client/cache/listers_extensions.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// If a deployment with a nil or empty selector creeps in, it should match nothing, not everything.
+		if selector.Empty() || !selector.Matches(labels.Set(pod.Labels)) {
+			continue
+		}
+		deployments = append(deployments, d)
+	}
+	if len(deployments) == 0 {
+		err = fmt.Errorf("could not find deployments set for Deployment %s in namespace %s with labels: %v", pod.Name, pod.Namespace, pod.Labels)
```
Comment (ncdc) [inline pkg/client/cache/listers_extensions.go]: Deployment -> Pod

File: pkg/controller/deployment/deployment_controller.go [hunk 5]
```diff
@@ -67,27 +67,27 @@ type DeploymentController struct {
 	syncHandler func(dKey string) error
 
 	// A store of deployments, populated by the dController
-	dStore cache.StoreToDeploymentLister
+	dLister cache.StoreToDeploymentLister
 	// Watches changes to all deployments
 	dController *cache.Controller
 	// A store of ReplicaSets, populated by the rsController
-	rsStore cache.StoreToReplicaSetLister
+	rsLister cache.StoreToReplicaSetLister
```
Comment (ncdc) [inline pkg/controller/deployment/deployment_controller.go]: I assume the rs changes in this commit should really be in the rs commit, but I'm not going to make you change them 😄

File: pkg/controller/informers/extensions.go [hunk 6]
```diff
@@ -68,3 +68,46 @@ func (f *daemonSetInformer) Lister() *cache.StoreToDaemonSetLister {
 	informer := f.Informer()
 	return &cache.StoreToDaemonSetLister{Store: informer.GetIndexer()}
 }
+
+// Deployment is type of SharedIndexInformer which watches and lists all pods.
```
Comment (ncdc) [inline pkg/controller/informers/extensions.go]: is a type

File: pkg/controller/informers/extensions.go [hunk 7]
```diff
@@ -68,3 +68,46 @@ func (f *daemonSetInformer) Lister() *cache.StoreToDaemonSetLister {
...
+// Interface provides constructor for informer and lister for pods
```
Comment (ncdc) [inline pkg/controller/informers/extensions.go]: Did you intend to write "Interface" here, as in "the interface", or did you mean "Deployment"?

File: pkg/controller/informers/extensions.go [hunk 8]
```diff
@@ -68,3 +68,46 @@ func (f *daemonSetInformer) Lister() *cache.StoreToDaemonSetLister {
...
+type Deployment interface {
+	Informer() cache.SharedIndexInformer
```
Comment (ncdc) [inline pkg/controller/informers/extensions.go]: Godoc?

**Review by ncdc — CHANGES_REQUESTED** [review]

**Comment by kargakis** [comment]
cc: @mfojtik 

Ref: https://github.com/kubernetes/kubernetes/issues/31780

File: pkg/controller/informers/extensions.go [hunk 9]
```diff
@@ -71,7 +71,7 @@ func (f *daemonSetInformer) Lister() *cache.StoreToDaemonSetLister {
 
 // Deployment is type of SharedIndexInformer which watches and lists all pods.
```
Comment (ncdc) [inline pkg/controller/informers/extensions.go]: Deployment->DeploymentInformer

**Review by ncdc — CHANGES_REQUESTED** [review]

File: pkg/client/cache/listers_extensions.go (hunk 1 above)
Comment (deads2k) [inline pkg/client/cache/listers_extensions.go]: > You didn't migrate Exists. Change comment to only mention List?

Looks I caught it later.

File: pkg/client/cache/listers_extensions.go (hunk 2 above)
Comment (deads2k) [inline pkg/client/cache/listers_extensions.go]: > There is some godoc for this type here if you want it: 25f42a2#diff-53f28720440ab4b589ef2273346cecfcL197

I don't think it adds much, do you?  It's boilerplate on all of these.

File: pkg/client/cache/listers_extensions.go (hunk 3 above)
Comment (deads2k) [inline pkg/client/cache/listers_extensions.go]: > The original error was "no deployments found for Pod" - change it back?

done

File: pkg/client/cache/listers_extensions.go (hunk 4 above)
Comment (deads2k) [inline pkg/client/cache/listers_extensions.go]: > Deployment -> Pod

done

File: pkg/controller/informers/extensions.go (hunk 8 above)
Comment (deads2k) [inline pkg/controller/informers/extensions.go]: > Godoc?

boilerplate

File: pkg/client/cache/listers_extensions.go (hunk 1 above)
Comment (ncdc) [inline pkg/client/cache/listers_extensions.go]: Yup

File: pkg/controller/informers/extensions.go (hunk 8 above)
Comment (ncdc) [inline pkg/controller/informers/extensions.go]: Fine to omit

**Comment by deads2k** [comment]
comments addressed.

**Comment by deads2k** [comment]
@ncdc did you make it through?

**Comment by ncdc** [comment]
@deads2k I did. It looks like there are 2 small outstanding typos.

**Comment by deads2k** [comment]
> @deads2k I did. It looks like there are 2 small outstanding typos.

Got them.

**Review by ncdc — APPROVED** [review]

**Pull request merged**